File: dnn-providers/hip-kernel-provider/rocke/platform/tests/instances/parity/gfx1250_lowering_emit.c
/* Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
 * SPDX-License-Identifier: MIT
 *
 * tests/parity/gfx1250_lowering_emit.c -- C-side emitter for the gfx1250
 * lowering surface. Builds each kernel identically to gfx1250_lowering_emit.py
 * so run_diff.py can byte-compare the two engines' .ll; see that file for the
 * config-by-config rationale and for why the gfx950 twins are included.
 *
 * arch is per-config (see CONFIGS), flavor = AUTO (matches the Python side).
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "rocke/ir.h"
#include "rocke/ir_serialize.h"
#include "rocke/lower_llvm.h"
#include "rocke/verify.h"

/* Every builder call below is its own statement, and the statement order
 * matches the Python emitter exactly. Nesting a builder call inside an argument
 * list would leave the order to the C compiler (argument evaluation order is
 * unspecified), and since each call consumes a value id, a different order
 * emits the same IR under different SSA names -- a byte mismatch with no real
 * defect behind it. */

static rocke_value_t*
    frag_param(rocke_ir_builder_t* b, const char* name, const rocke_type_t* elem, bool readonly)
{
    rocke_param_opts_t o;
    memset(&o, 0, sizeof(o));
    o.noalias = true;
    o.noalias_set = true;
    if(readonly)
    {
        o.readonly = true;
        o.readonly_set = true;
    }
    o.align = 16;
    o.align_set = true;
    return rocke_b_param(b, name, rocke_ptr_type(b, elem, "global"), &o);
}

/* Mirrors the Python _frag_operands helper: A and B as <n x elem>, C as
 * <8 x float>, all loaded so the operands stay opaque to the lowerer. */
typedef struct frag_ops
{
    rocke_value_t* tid;
    rocke_value_t* c_ptr;
    rocke_value_t* a;
    rocke_value_t* b;
    rocke_value_t* c;
} frag_ops_t;

static frag_ops_t frag_operands(rocke_ir_builder_t* b, const rocke_type_t* elem, int n)
{
    frag_ops_t f;
    rocke_value_t* a_ptr = frag_param(b, "A", elem, true);
    rocke_value_t* b_ptr = frag_param(b, "B", elem, true);
    f.c_ptr = frag_param(b, "C", rocke_f32(), false);
    f.tid = rocke_b_thread_id_x(b);
    f.a = rocke_b_global_load_vN(b, a_ptr, f.tid, elem, n, /*align=*/0);
    f.b = rocke_b_global_load_vN(b, b_ptr, f.tid, elem, n, /*align=*/0);
    f.c = rocke_b_global_load_vN(b, f.c_ptr, f.tid, rocke_f32(), 8, /*align=*/0);
    return f;
}

/* K=32 f16/bf16 WMMA: the gfx1250 8-operand signature. bf16 is the interesting
 * half -- gfx11/gfx12 bitcast the operands to <16 x i16> before the call, while
 * gfx1250 takes <16 x bfloat> directly. */
static void wmma_k32(rocke_ir_builder_t* b, const rocke_type_t* elem, const char* op_id)
{
    frag_ops_t f = frag_operands(b, elem, 16);
    rocke_value_t* d = rocke_b_mma(b, op_id, f.a, f.b, f.c, NULL, 0);
    rocke_b_global_store(b, f.c_ptr, f.tid, d, /*align=*/1);
    rocke_b_ret(b);
}

static void build_wmma_k32_f16(rocke_ir_builder_t* b)
{
    wmma_k32(b, rocke_f16(), "wmma_gfx1250_f32_16x16x32_f16");
}

static void build_wmma_k32_bf16(rocke_ir_builder_t* b)
{
    wmma_k32(b, rocke_bf16(), "wmma_gfx1250_f32_16x16x32_bf16");
}

/* K=64 fp8/bf8 WMMA: the gfx1250 6-operand signature. The fragments arrive as
 * <8 x i32> (32 packed bytes per lane), so the dtype pair lives only in the
 * op_id and the mangled intrinsic name. */
static void wmma_k64(rocke_ir_builder_t* b, const char* op_id)
{
    frag_ops_t f = frag_operands(b, rocke_i32(), 8);
    rocke_value_t* d = rocke_b_mma(b, op_id, f.a, f.b, f.c, NULL, 0);
    rocke_b_global_store(b, f.c_ptr, f.tid, d, /*align=*/1);
    rocke_b_ret(b);
}

static void build_wmma_k64_fp8_fp8(rocke_ir_builder_t* b)
{
    wmma_k64(b, "wmma_gfx1250_f32_16x16x64_fp8_fp8");
}

static void build_wmma_k64_fp8_bf8(rocke_ir_builder_t* b)
{
    wmma_k64(b, "wmma_gfx1250_f32_16x16x64_fp8_bf8");
}

static void build_wmma_k64_bf8_fp8(rocke_ir_builder_t* b)
{
    wmma_k64(b, "wmma_gfx1250_f32_16x16x64_bf8_fp8");
}

static void build_wmma_k64_bf8_bf8(rocke_ir_builder_t* b)
{
    wmma_k64(b, "wmma_gfx1250_f32_16x16x64_bf8_bf8");
}

/* K=128 FP8 SCALE/SCALE16 WMMA. Matrix fragments are <16 x i32>; packed
 * E8M0 scale operands are i32 for SCALE and i64 for SCALE16. */
static void wmma_scaled(rocke_ir_builder_t* b, bool scale16)
{
    const rocke_type_t* scale_ty = scale16 ? rocke_i64() : rocke_i32();
    rocke_value_t* a_ptr = frag_param(b, "A", rocke_i32(), true);
    rocke_value_t* b_ptr = frag_param(b, "B", rocke_i32(), true);
    rocke_value_t* c_ptr = frag_param(b, "C", rocke_f32(), false);
    rocke_value_t* scale_ptr = frag_param(b, "scale", scale_ty, true);
    rocke_value_t* tid = rocke_b_thread_id_x(b);
    rocke_value_t* a_lo = rocke_b_global_load_vN(b, a_ptr, tid, rocke_i32(), 8, /*align=*/0);
    rocke_value_t* eight = rocke_b_const_i32(b, 8);
    rocke_value_t* hi_idx = rocke_b_add(b, tid, eight);
    rocke_value_t* a_hi = rocke_b_global_load_vN(b, a_ptr, hi_idx, rocke_i32(), 8, /*align=*/0);
    rocke_value_t* a = rocke_b_vec_concat(b, a_lo, a_hi);
    rocke_value_t* b_lo = rocke_b_global_load_vN(b, b_ptr, tid, rocke_i32(), 8, /*align=*/0);
    rocke_value_t* b_hi = rocke_b_global_load_vN(b, b_ptr, hi_idx, rocke_i32(), 8, /*align=*/0);
    rocke_value_t* bb = rocke_b_vec_concat(b, b_lo, b_hi);
    rocke_value_t* c = rocke_b_global_load_vN(b, c_ptr, tid, rocke_f32(), 8, /*align=*/0);
    rocke_value_t* scale = rocke_b_global_load(b, scale_ptr, tid, scale_ty, /*align=*/1);
    rocke_value_t* d = scale16
                           ? rocke_b_wmma_scale16_f32_16x16x128_fp8_fp8(b, a, bb, c, scale, scale)
                           : rocke_b_wmma_scale_f32_16x16x128_fp8_fp8(b, a, bb, c, scale, scale);
    rocke_b_global_store(b, c_ptr, tid, d, /*align=*/1);
    rocke_b_ret(b);
}

static void build_wmma_scale(rocke_ir_builder_t* b)
{
    wmma_scaled(b, false);
}

static void build_wmma_scale16(rocke_ir_builder_t* b)
{
    wmma_scaled(b, true);
}

/* ds_read_b128_tr_b16. gfx950 has one type-agnostic opcode returning
 * <8 x i16> that the handler reinterprets; gfx1250 has per-element-type
 * opcodes (.v8f16 / .v8bf16) that land in the right type with no reinterpret. */
static void tr16_b128(rocke_ir_builder_t* b, const rocke_type_t* elem)
{
    rocke_param_opts_t o;
    const int shape[] = {64, 8};
    rocke_value_t* out;
    rocke_value_t* smem;
    rocke_value_t* tid;
    rocke_value_t* zero;
    rocke_value_t* v;
    rocke_value_t* idx[2];

    memset(&o, 0, sizeof(o));
    o.noalias = true;
    o.noalias_set = true;
    o.align = 16;
    o.align_set = true;
    out = rocke_b_param(b, "out", rocke_ptr_type(b, elem, "global"), &o);
    smem = rocke_b_smem_alloc(b, elem, shape, 2, "tile");
    tid = rocke_b_thread_id_x(b);
    zero = rocke_b_const_i32(b, 0);
    idx[0] = tid;
    idx[1] = zero;
    v = rocke_b_ds_read_tr16_b128(b, smem, idx, 2, elem);
    rocke_b_global_store(b, out, tid, v, /*align=*/1);
    rocke_b_ret(b);
}

static void build_tr16_f16(rocke_ir_builder_t* b)
{
    tr16_b128(b, rocke_f16());
}

static void build_tr16_bf16(rocke_ir_builder_t* b)
{
    tr16_b128(b, rocke_bf16());
}

/* The two LDS workgroup barriers. An s_barrier does not drain outstanding LDS
 * traffic, so each barrier is preceded by a wait: gfx9/10/11 spend one
 * monolithic s_waitcnt, gfx1250 emits split s_wait_loadcnt / s_wait_dscnt (and
 * for the LDS-only barrier, dscnt alone). */
static void build_barrier_drains(rocke_ir_builder_t* b)
{
    rocke_param_opts_t o;
    const int shape[] = {64, 8};
    rocke_value_t* out;
    rocke_value_t* smem;
    rocke_value_t* tid;
    rocke_value_t* zero;
    rocke_value_t* v;
    rocke_value_t* idx[2];

    memset(&o, 0, sizeof(o));
    o.noalias = true;
    o.noalias_set = true;
    o.align = 16;
    o.align_set = true;
    out = rocke_b_param(b, "out", rocke_ptr_type(b, rocke_f16(), "global"), &o);
    smem = rocke_b_smem_alloc(b, rocke_f16(), shape, 2, "tile");
    tid = rocke_b_thread_id_x(b);
    zero = rocke_b_const_i32(b, 0);
    rocke_b_sync(b);
    idx[0] = tid;
    idx[1] = zero;
    v = rocke_b_smem_load_vN_f16(b, smem, idx, 2, 8);
    rocke_b_sync_lds_only(b);
    rocke_b_global_store(b, out, tid, v, /*align=*/1);
    rocke_b_ret(b);
}

/* The two wait-counter facts: s_wait_asynccnt lowers to nothing without an
 * async-DMA counter and to the intrinsic on gfx1250; s_waitcnt is the mirror
 * image, since llvm.amdgcn.s.waitcnt is not selectable on gfx1250. */
static void build_wait_counters(rocke_ir_builder_t* b)
{
    rocke_b_s_wait_asynccnt(b, 0);
    rocke_b_s_waitcnt(b, /*vmcnt=*/0, /*lgkmcnt=*/0, /*expcnt=*/-1);
    rocke_b_s_wait_asynccnt(b, 3);
    rocke_b_s_waitcnt(b, /*vmcnt=*/-1, /*lgkmcnt=*/0, /*expcnt=*/-1);
    rocke_b_ret(b);
}

static void build_standalone_controls(rocke_ir_builder_t* b)
{
    const int shape[] = {1};
    rocke_value_t* barrier = rocke_b_smem_alloc(b, rocke_i64(), shape, 1, "named_barrier");
    rocke_value_t* barrier_ptr = rocke_b_smem_addr_of(b, barrier);
    rocke_value_t* members = rocke_b_const_i32(b, 2);
    rocke_b_s_wait_tensorcnt(b, 3);
    rocke_b_s_barrier_signal(b, 1);
    rocke_b_s_barrier_wait(b, 1);
    rocke_b_s_barrier_init(b, barrier_ptr, members);
    rocke_b_s_barrier_signal_var(b, barrier_ptr, members);
    rocke_b_s_barrier_join(b, barrier_ptr);
    rocke_b_s_wakeup_barrier(b, barrier_ptr);
    rocke_b_s_barrier_leave(b, 1);
    rocke_b_s_delay_alu(b, 0x1234);
    rocke_b_s_wait_alu(b, 0x2345);
    rocke_b_s_clause(b, 0x3456);
    rocke_b_s_wait_xcnt(b, 0x4567);
    rocke_b_ret(b);
}

static void build_async_store(rocke_ir_builder_t* b)
{
    const int shape[] = {4};
    rocke_param_opts_t o;
    memset(&o, 0, sizeof(o));
    o.noalias = true;
    o.noalias_set = true;
    o.align = 16;
    o.align_set = true;
    rocke_value_t* out = rocke_b_param(b, "out", rocke_ptr_type(b, rocke_i32(), "global"), &o);
    rocke_value_t* smem = rocke_b_smem_alloc(b, rocke_i32(), shape, 1, "store_src");
    rocke_value_t* lds_ptr = rocke_b_smem_addr_of(b, smem);
    const int widths[] = {1, 4, 8, 16};
    for(int i = 0; i < 4; ++i)
        rocke_b_global_store_async_from_lds(
            b, out, lds_ptr, widths[i], widths[i], /*cachepolicy=*/3);
    rocke_b_ret(b);
}

static void global_tr16(rocke_ir_builder_t* b, const rocke_type_t* elem)
{
    rocke_param_opts_t src_o;
    rocke_param_opts_t out_o;
    memset(&src_o, 0, sizeof(src_o));
    memset(&out_o, 0, sizeof(out_o));
    src_o.noalias = true;
    src_o.noalias_set = true;
    src_o.readonly = true;
    src_o.readonly_set = true;
    src_o.align = 16;
    src_o.align_set = true;
    out_o.noalias = true;
    out_o.noalias_set = true;
    out_o.align = 16;
    out_o.align_set = true;
    rocke_value_t* src = rocke_b_param(b, "src", rocke_ptr_type(b, elem, "global"), &src_o);
    rocke_value_t* out = rocke_b_param(b, "out", rocke_ptr_type(b, elem, "global"), &out_o);
    rocke_value_t* zero = rocke_b_const_i32(b, 0);
    rocke_value_t* value = rocke_b_global_load_tr16_b128(b, src, elem);
    rocke_b_global_store(b, out, zero, value, /*align=*/1);
    rocke_b_ret(b);
}

static void build_global_tr16_f16(rocke_ir_builder_t* b)
{
    global_tr16(b, rocke_f16());
}

static void build_global_tr16_bf16(rocke_ir_builder_t* b)
{
    global_tr16(b, rocke_bf16());
}

static void build_global_tr16_i16(rocke_ir_builder_t* b)
{
    global_tr16(b, rocke_i16());
}

static void build_tensor_transfers(rocke_ir_builder_t* b)
{
    rocke_value_t* d4 = rocke_b_zero_vec(b, rocke_i32(), 4);
    rocke_value_t* d8 = rocke_b_zero_vec(b, rocke_i32(), 8);
    rocke_b_tensor_load_to_lds(b, d4, d8, d4, d4, d8, 5);
    rocke_b_tensor_store_from_lds(b, d4, d8, d4, d4, d8, 5);
    rocke_b_s_wait_tensorcnt(b, 0);
    rocke_b_ret(b);
}

typedef void (*build_fn_t)(rocke_ir_builder_t*);

typedef struct config
{
    build_fn_t build;
    const char* arch;
} config_t;

/* Each gfx1250 config that tests a *choice* of encoding is followed by its
 * gfx950 twin, so the pair pins both branches. Index-for-index with the Python
 * emitter's CONFIGS list. */
static const config_t CONFIGS[] = {
    {build_wmma_k32_f16, "gfx1250"},
    {build_wmma_k32_bf16, "gfx1250"},
    {build_wmma_k64_fp8_fp8, "gfx1250"},
    {build_wmma_k64_fp8_bf8, "gfx1250"},
    {build_wmma_k64_bf8_fp8, "gfx1250"},
    {build_wmma_k64_bf8_bf8, "gfx1250"},
    {build_wmma_scale, "gfx1250"},
    {build_wmma_scale16, "gfx1250"},
    {build_tr16_f16, "gfx1250"},
    {build_tr16_f16, "gfx950"},
    {build_tr16_bf16, "gfx1250"},
    {build_tr16_bf16, "gfx950"},
    {build_barrier_drains, "gfx1250"},
    {build_barrier_drains, "gfx950"},
    {build_wait_counters, "gfx1250"},
    {build_wait_counters, "gfx950"},
    {build_standalone_controls, "gfx1250"},
    {build_async_store, "gfx1250"},
    {build_global_tr16_f16, "gfx1250"},
    {build_global_tr16_bf16, "gfx1250"},
    {build_global_tr16_i16, "gfx1250"},
    {build_tensor_transfers, "gfx1250"},
};

static const int NUM_CONFIGS = (int)(sizeof(CONFIGS) / sizeof(CONFIGS[0]));

int main(int argc, char** argv)
{
    if(argc < 2)
    {
        fprintf(
            stderr, "usage: %s <config_index 0..%d> [ll|ir|verify]\n", argv[0], NUM_CONFIGS - 1);
        return 2;
    }
    int idx = atoi(argv[1]);
    const char* mode = (argc > 2) ? argv[2] : "ll";

    if(strcmp(mode, "ll") != 0 && strcmp(mode, "ir") != 0 && strcmp(mode, "verify") != 0)
    {
        fprintf(stderr, "unknown mode %s\n", mode);
        return 2;
    }
    if(idx < 0 || idx >= NUM_CONFIGS)
    {
        fprintf(stderr, "unknown config index %d\n", idx);
        return 2;
    }

    rocke_ir_builder_t b;
    if(rocke_ir_builder_init(&b, "gfx1250_lowering") != ROCKE_OK)
    {
        fprintf(stderr, "builder init failed\n");
        return 1;
    }
    /* Python: b.kernel.attrs["max_workgroup_size"] = 64 */
    rocke_attr_set_int(&b, &b.kernel->attrs, "max_workgroup_size", 64);
    CONFIGS[idx].build(&b);

    if(!rocke_ir_builder_ok(&b))
    {
        fprintf(stderr, "builder error: %s\n", rocke_ir_builder_error(&b));
        rocke_ir_builder_free(&b);
        return 1;
    }

    rocke_kernel_def_t* kernel = rocke_ir_builder_kernel(&b);
    if(strcmp(mode, "ll") == 0)
    {
        char* llvm_text = NULL;
        char err[ROCKE_ERR_MSG_CAP];
        err[0] = 0;
        rocke_status_t st = rocke_lower_kernel_to_llvm_ex(
            kernel, ROCKE_LLVM_FLAVOR_AUTO, CONFIGS[idx].arch, &llvm_text, err, sizeof err);
        if(st != ROCKE_OK || !llvm_text)
        {
            fprintf(stderr, "lower failed: status=%d err=%s\n", (int)st, err);
            rocke_ir_builder_free(&b);
            return 1;
        }
        fputs(llvm_text, stdout);
        free(llvm_text);
    }
    else if(strcmp(mode, "ir") == 0)
    {
        char* text = NULL;
        rocke_status_t st = rocke_ir_serialize(kernel, &text);
        if(st != ROCKE_OK || !text)
        {
            fprintf(stderr, "serialize failed: status=%d\n", (int)st);
            rocke_ir_builder_free(&b);
            return 1;
        }
        fputs(text, stdout);
        free(text);
    }
    else
    { /* verify */
        rocke_diag_t* d = NULL;
        size_t n = 0;
        rocke_verify(kernel, &d, &n);
        for(size_t i = 0; i < n; i++)
        {
            char* s = rocke_diag_to_string(&d[i]);
            if(s)
            {
                puts(s);
                free(s);
            }
        }
        rocke_diags_free(d, n);
    }

    rocke_ir_builder_free(&b);
    return 0;
}

File: dnn-providers/hip-kernel-provider/rocke/platform/tests/instances/parity/gfx1250_lowering_emit.py
#!/usr/bin/env python3
# Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
# SPDX-License-Identifier: MIT
#
# tests/parity/gfx1250_lowering_emit.py -- Python reference emitter for the
# gfx1250 lowering surface.
#
# gfx1250 kernels are authored in Python (there is no cpp/instances/gfx1250/),
# so no instance-builder family in this gate exercises the gfx1250 *lowerer*.
# That left the target-specific places Gfx1250Backend diverges from its Gfx12RdnaBackend
# parent covered only by the ROCKE_BACKEND=both pytest lane, and invisible to
# check_byte_identity.py -- the tree's stated definition-of-done. This family
# closes that: one config per divergence, built from the smallest kernel that
# reaches it, byte-compared against gfx1250_lowering_emit.c.
#
# The gfx950 twins are deliberately included. Several divergences are a
# *choice between two encodings*, so a lowering that ignored the backend and
# always picked the gfx1250 form would still pass a gfx1250-only family. Pairing
# each with its gfx950 counterpart pins both sides of the choice.
#
# arch is per-config (see _spec), llvm_flavor = AUTO, matching the C side.
from rocke.core.ir import BF16, F16, F32, I16, I32, I64, IRBuilder, KernelDef, PtrType

from _emit_common import run_emit


def _frag_operands(b, elem, n):
    """Load the three MMA fragments: A and B as <n x elem>, C as <8 x float>.

    Loading them (rather than materializing constants) keeps the operands
    opaque to the lowerer, so the WMMA call site is the only thing the config
    is testing.
    """
    a_ptr = b.param("A", PtrType(elem, "global"), noalias=True, readonly=True, align=16)
    b_ptr = b.param("B", PtrType(elem, "global"), noalias=True, readonly=True, align=16)
    c_ptr = b.param("C", PtrType(F32, "global"), noalias=True, align=16)
    tid = b.thread_id_x()
    a = b.global_load_vN(a_ptr, tid, dtype=elem, n=n)
    bb = b.global_load_vN(b_ptr, tid, dtype=elem, n=n)
    c = b.global_load_vN(c_ptr, tid, dtype=F32, n=8)
    return tid, c_ptr, a, bb, c


def _wmma_k32(elem):
    """K=32 f16/bf16 WMMA: the gfx1250 8-operand signature.

    bf16 is the interesting half: gfx11/gfx12 bitcast the operands to
    <16 x i16> before the call, while gfx1250 takes <16 x bfloat> directly.
    """

    def build(b: IRBuilder) -> None:
        tid, c_ptr, a, bb, c = _frag_operands(b, elem, 16)
        suffix = "f16" if elem is F16 else "bf16"
        d = b.mma(f"wmma_gfx1250_f32_16x16x32_{suffix}", a, bb, c)
        b.global_store(c_ptr, tid, d)
        b.ret()

    return build


def _wmma_k64(a_kind, b_kind):
    """K=64 fp8/bf8 WMMA: the gfx1250 6-operand signature.

    The fragments arrive as <8 x i32> (32 packed bytes per lane), so the
    dtype pair lives only in the op_id and the mangled intrinsic name.
    """

    def build(b: IRBuilder) -> None:
        tid, c_ptr, a, bb, c = _frag_operands(b, I32, 8)
        d = b.mma(f"wmma_gfx1250_f32_16x16x64_{a_kind}_{b_kind}", a, bb, c)
        b.global_store(c_ptr, tid, d)
        b.ret()

    return build


def _wmma_scaled(a_kind, b_kind, scale_mode):
    """K=128 scaled WMMA, parameterized by operand dtypes and scale mode."""
    scale_ty = {"scale": I32, "scale16": I64}[scale_mode]
    op_id = f"wmma_{scale_mode}_f32_16x16x128_{a_kind}_{b_kind}"

    def build(b: IRBuilder) -> None:
        a_ptr = b.param(
            "A", PtrType(I32, "global"), noalias=True, readonly=True, align=16
        )
        b_ptr = b.param(
            "B", PtrType(I32, "global"), noalias=True, readonly=True, align=16
        )
        c_ptr = b.param("C", PtrType(F32, "global"), noalias=True, align=16)
        scale_ptr = b.param(
            "scale",
            PtrType(scale_ty, "global"),
            noalias=True,
            readonly=True,
            align=16,
        )
        tid = b.thread_id_x()
        a_lo = b.global_load_vN(a_ptr, tid, dtype=I32, n=8)
        eight = b.const_i32(8)
        hi_idx = b.add(tid, eight)
        a_hi = b.global_load_vN(a_ptr, hi_idx, dtype=I32, n=8)
        a = b.vec_concat(a_lo, a_hi)
        b_lo = b.global_load_vN(b_ptr, tid, dtype=I32, n=8)
        b_hi = b.global_load_vN(b_ptr, hi_idx, dtype=I32, n=8)
        bb = b.vec_concat(b_lo, b_hi)
        c = b.global_load_vN(c_ptr, tid, dtype=F32, n=8)
        scale = b.global_load(scale_ptr, tid, dtype=scale_ty)
        d = b.mma(op_id, a, bb, c, scale, scale)
        b.global_store(c_ptr, tid, d)
        b.ret()

    return build


def _tr16_b128(elem):
    """ds_read_b128_tr_b16.

    gfx950 has one type-agnostic opcode returning <8 x i16> that the handler
    reinterprets; gfx1250 has per-element-type opcodes (.v8f16 / .v8bf16) that
    land in the right type with no reinterpret. Run on both arches so the
    element-typed selection cannot be hardcoded either way.
    """

    def build(b: IRBuilder) -> None:
        out = b.param("out", PtrType(elem, "global"), noalias=True, align=16)
        smem = b.smem_alloc(elem, [64, 8], name_hint="tile")
        tid = b.thread_id_x()
        zero = b.const_i32(0)
        v = b.ds_read_tr16_b128(smem, tid, zero, dtype=elem)
        b.global_store(out, tid, v)
        b.ret()

    return build


def build_barrier_drains(b: IRBuilder) -> None:
    """The two LDS workgroup barriers.

    An s_barrier does not drain outstanding LDS traffic, so each barrier is
    preceded by a wait. gfx9/10/11 spend one monolithic s_waitcnt on it;
    gfx1250 emits split s_wait_loadcnt / s_wait_dscnt (and for the LDS-only
    barrier, dscnt alone -- the VMEM chain deliberately stays in flight).
    Both barriers are here so the drain_vmem=True and =False paths are
    distinguished.
    """
    out = b.param("out", PtrType(F16, "global"), noalias=True, align=16)
    smem = b.smem_alloc(F16, [64, 8], name_hint="tile")
    tid = b.thread_id_x()
    zero = b.const_i32(0)
    b.sync()
    v = b.smem_load_vN(smem, tid, zero, dtype=F16, n=8)
    b.sync_lds_only()
    b.global_store(out, tid, v)
    b.ret()


def build_wait_counters(b: IRBuilder) -> None:
    """The two wait-counter facts.

    ``s_wait_asynccnt`` lowers to nothing on a backend with no async-DMA
    counter and to the intrinsic on gfx1250; ``s_waitcnt`` is the mirror
    image, since llvm.amdgcn.s.waitcnt is not selectable on gfx1250. A config
    that emitted both, or neither, on one arch would be wrong on the other.
    """
    b.s_wait_asynccnt(0)
    b.s_waitcnt(vmcnt=0, lgkmcnt=0)
    b.s_wait_asynccnt(3)
    b.s_waitcnt(lgkmcnt=0)
    b.ret()


def build_standalone_controls(b: IRBuilder) -> None:
    barrier = b.smem_alloc(I64, [1], name_hint="named_barrier")
    barrier_ptr = b.smem_addr_of(barrier)
    members = b.const_i32(2)
    b.s_wait_tensorcnt(3)
    b.s_barrier_signal(1)
    b.s_barrier_wait(1)
    b.s_barrier_init(barrier_ptr, members)
    b.s_barrier_signal_var(barrier_ptr, members)
    b.s_barrier_join(barrier_ptr)
    b.s_wakeup_barrier(barrier_ptr)
    b.s_barrier_leave(1)
    b.s_delay_alu(0x1234)
    b.s_wait_alu(0x2345)
    b.s_clause(0x3456)
    b.s_wait_xcnt(0x4567)
    b.ret()


def build_async_store(b: IRBuilder) -> None:
    out = b.param("out", PtrType(I32, "global"), noalias=True, align=16)
    smem = b.smem_alloc(I32, [4], name_hint="store_src")
    lds_ptr = b.smem_addr_of(smem)
    for width in (1, 4, 8, 16):
        b.global_store_async_from_lds(
            out,
            lds_ptr,
            width_bytes=width,
            offset_bytes=width,
            cachepolicy=3,
        )
    b.ret()


def _global_tr16(elem):
    def build(b: IRBuilder) -> None:
        src = b.param(
            "src", PtrType(elem, "global"), noalias=True, readonly=True, align=16
        )
        out = b.param("out", PtrType(elem, "global"), noalias=True, align=16)
        zero = b.const_i32(0)
        value = b.global_load_tr16_b128(src, dtype=elem)
        b.global_store(out, zero, value)
        b.ret()

    return build


def build_tensor_transfers(b: IRBuilder) -> None:
    d4 = b.zero_vec(I32, 4)
    d8 = b.zero_vec(I32, 8)
    b.tensor_load_to_lds(d4, d8, d4, d4, d8, cachepolicy=5)
    b.tensor_store_from_lds(d4, d8, d4, d4, d8, cachepolicy=5)
    b.s_wait_tensorcnt(0)
    b.ret()


# (builder, arch). Each gfx1250 config that tests a *choice* of encoding is
# followed by its gfx950 twin, so the pair pins both branches.
CONFIGS = [
    (_wmma_k32(F16), "gfx1250"),
    (_wmma_k32(BF16), "gfx1250"),
    (_wmma_k64("fp8", "fp8"), "gfx1250"),
    (_wmma_k64("fp8", "bf8"), "gfx1250"),
    (_wmma_k64("bf8", "fp8"), "gfx1250"),
    (_wmma_k64("bf8", "bf8"), "gfx1250"),
    (_wmma_scaled("fp8", "fp8", "scale"), "gfx1250"),
    (_wmma_scaled("fp8", "fp8", "scale16"), "gfx1250"),
    (_tr16_b128(F16), "gfx1250"),
    (_tr16_b128(F16), "gfx950"),
    (_tr16_b128(BF16), "gfx1250"),
    (_tr16_b128(BF16), "gfx950"),
    (build_barrier_drains, "gfx1250"),
    (build_barrier_drains, "gfx950"),
    (build_wait_counters, "gfx1250"),
    (build_wait_counters, "gfx950"),
    (build_standalone_controls, "gfx1250"),
    (build_async_store, "gfx1250"),
    (_global_tr16(F16), "gfx1250"),
    (_global_tr16(BF16), "gfx1250"),
    (_global_tr16(I16), "gfx1250"),
    (build_tensor_transfers, "gfx1250"),
]


def _spec(idx: int):
    """Config selector: the (builder, arch) pair the shared driver expects."""
    if not 0 <= idx < len(CONFIGS):
        raise SystemExit(f"unknown config index {idx}")
    return CONFIGS[idx]


def _build(build_fn, *, arch: str = "gfx1250") -> KernelDef:
    b = IRBuilder("gfx1250_lowering")
    b.kernel.attrs["max_workgroup_size"] = 64
    build_fn(b)
    return b.kernel


def main() -> int:
    return run_emit(
        _spec,
        _build,
        usage="usage: gfx1250_lowering_emit.py <config_index> [ll|ir|verify]\n",
        arch="gfx1250",
    )


if __name__ == "__main__":
    raise SystemExit(main())
